### PR TITLE
Embedding d3-cloud library + debounced renders

### PR DESCRIPTION
**Pull request by WhoAteDaCake**
This addresses the issues discussed in #30 .
Please let me know if this fits with the coding guidelines, or if there any things you'd like me to change

### CODE REVIEW

**Comment by chrisrzhou** [comment]
@WhoAteDaCake, awesome stuff! Would you mind updating the PR summary with some links/resources that inspire your approach here?

It looks like we're moving away from `d3-cloud` and using a custom logic, and I would like to get more detail on it before making the switch.  It's good that this will fix the issue, but not relying on a maintained library would mean we (definitely myself :P) would need some knowledge/context on the implementation.

Thank you! Will review the PR this weekend and get back to you if you can update this PR with some context/links that will provide that knowledge to anyone else who comes across it in the future.

**Comment by WhoAteDaCake** [comment]
Hi, thanks for taking a look. So in the PR itself, I didn't add much new code, except taking the algorithm from d3-cloud repo and splitting it out in to separate files. To be honest I am not 100% sure how the algorithm itself works. I could link where specific parts come from. Would that help, or should we take a different approach

**Comment by chrisrzhou** [comment]
👍  Ah I see, l think it's important to rely on `d3-cloud`'s source code as a dependency instead of managing it as separate files ourselves.  Was there a reason why you decided to split the files?

Ideally, we would revert the split files and use `d3-cloud` and keep the PR only focused on code changes relating to debounced rendering.

**Comment by WhoAteDaCake** [comment]
Hi, so the reason why I've embedded the algorithm was because I had to change the loop used to calculate the word positions. As you can see `d3-cloud` uses a [while loop](https://github.com/jasondavies/d3-cloud/blob/master/index.js#L56), which is can cause problems when rendering large datasets or multiple worldclouds, as It blocks the main thread and can make the page [not responsive](https://codesandbox.io/s/staging-forest-8qkyi) for few seconds. I've changed it to use a `setTimeout` [loop](https://github.com/WhoAteDaCake/react-wordcloud/blob/cloud-integration/src/cloud/index.ts#L71) and it seems to perform better.

**Review by chrisrzhou — COMMENTED** [review]
@WhoAteDaCake, thank you for the detailed explanation and this makes a lot of sense.  Great approach and good job at isolating a performance fix for it.  A few high level things to organize:

- Let's import `lodash.debounce` as a dependency instead of implementing one on our own.  I know it's simple enough to implement it here, but we can rely on a tested and maintained library if any future issues crop up
- I really appreciate the hardwork in organizing the code in `helpers.ts`, `math.ts` and `cloud/index.ts`.  However, I would ask if you can put all methods in this file into just `cloud.js` and at the top of the file, leave documentation like this:
```js
/**
 * Implements the wordcloud algorithm based on: https://github.com/jasondavies/d3-cloud/tree/v1.2.5
 * Improves performance by applying a timeout to the cloud layout (lines ***-***)
 */

// leave appropriate comments in lines ***-***
```

Ideally, we want to have the mental model of "keeping the code as close to the source code" instead of "rewriting the code to fit this project".  The above will help future maintainers and users know what to look out for if they want to make an active decision about using this library.  In this spirit, I fully support "copy-pasting" the source code, leaving it non-typescripted, and adjusting minimal lines for the code to work.  This will provide us with a path for maintaining this library if it deviates from `d3-cloud` repo itself.

One last question: Have you considered on improving `d3-cloud` itself? I feel the work you've done here is great and would be great if you improved the underlying dependency itself!  I will let you figure that out yourself and we can refactor if that is done in the future.  For now, I think your fix here is a great addition to making this library more usable!

File: src/index.tsx [hunk 1]
```diff
@@ -66,6 +72,7 @@ export default function Wordcloud({
   options,
   size: initialSize,
   words,
+  renderDebounce = 100,
```
Comment (chrisrzhou) [inline src/index.tsx]: Let's integrate this in `options` prop.  At some point I had the choice to use flat props or organize configuration related props in an `options` object.  I chose the latter, so I think that's the right place for this to go in.  Don't worry too much about updating the documentation.  If you update the `propTypes` comment string, that should show up at least in the `Props` documentation page.

**Comment by WhoAteDaCake** [comment]
Hi, thanks for the feedback @chrisrzhou . 

> Have you considered on improving d3-cloud itself?

Yes eventually, it's just that I do not have as much free time right now and it might take a while. If the algorithm works, I am not quite sure how much more beneficial would it be to fully re-write it in typescript. I would be interested to take a look at it though. (I guess maintenance would be easier)

I've undone my changes and essentially copy-pasted the algorithm to a single file + added some comments. I've also copy pasted the type definition of `@types/d3-cloud` and added the new functionality. I am not sure if that's the best approach or should we keep relying on `@types/d3-cloud`.

**Comment by chrisrzhou** [comment]
This looks great! I'll test tonight and update the PR with other related changes.  Thank you so much for handling this and making the library better!

**Comment by WhoAteDaCake** [comment]
We use this at work, so I am just glad i was able to help make it better 🙂

File: src/cloud.ts [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+          // Temporary hack
+          d.x -= size[0] >> 1;
+          d.y -= size[1] >> 1;
+        }
+      }
+    }
+
+    // Added by react-wordcloud
+    // Iterates dataset using setTimeout in order to prevent blocking of the main thread
+    function loop(i: number) {
```
Comment (chrisrzhou) [inline src/cloud.ts]: Going through the PR to test the code and behaviors.  This function is defined and clearly used, but I am unable to trace it in the code (no mention of calling `loop` anywhere in our new code or in `d3-cloud`).

Any idea where this is being called?  I'll submit an update to this PR on some cleanup.

**Comment by chrisrzhou** [comment]
@WhoAteDaCake, I organized the code accordingly here: https://github.com/chrisrzhou/react-wordcloud/pull/35.  Documented the interfaces and split the code in the way I feel is best for future maintainence.

Let me know your thoughts.  There's still the layout bugs in #34 that I'm not sure how to address.  If you are able to test with this branch and fix the code accordingly, let me know!

**Comment by chrisrzhou** [comment]
@WhoAteDaCake, let me know if you are happy with the code that you've merged in (documentation and references).  Specifically, I made a note about how `enableOptimizations` is a 'beta' feature.  I don't feel completely confident to maintain this, and hope that the underlying `d3-cloud` library would address this.  Until then, I am creating this separation explicitly in the docs, but still supporting the feature because it's useful to consumers.

Feel free to address any other bugs you find.  I'll check back tonight and merge this to `master` once things are in place.

Thanks for all your awesome work on this!

**Comment by WhoAteDaCake** [comment]
Yeah, I am happy with the changes you made. It would be way better if this was implemented within d3-cloud codebase, so keeping this as a beta feature makes sense :).

**Comment by chrisrzhou** [comment]
Tested, works great! I'm checking other issues today, and will clean the code up a little bit before drafting a minor release.

**Pull request merged**